language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java

```diff
@@ -183,4 +201,121 @@ public static CodeActionNodeType topLevelNodeInLine(LSContext context, TextDocum
+                    if (child instanceof FunctionBodyBlockNode) {
+                        return createNodePositions(child.lineRange(), startCodePosition);
+                    }
+                }
+                break;
+            case "SERVICE_DECLARATION":
+                for (Node child : member.children()) {
+                    if (child instanceof ServiceBodyNode) {
+                        return createNodePositions(child.lineRange(), startCodePosition);
+                    }
+                }
+                break;
+            case "METHOD_DECLARATION":
+                LineRange methodLineRange = member.lineRange();
+                Position methodEndCodeBodyPosition = new Position(methodLineRange.endLine().line(),
+                        methodLineRange.endLine().offset());
+                return new NodePositions(startCodePosition, methodEndCodeBodyPosition, methodEndCodeBodyPosition);
+            case "CLASS_DEFINITION":
+            case "TYPE_DEFINITION":
+                Position startCodeBodyPosition = null;
+                for (Node child : member.children()) {
+                    if (child instanceof IdentifierToken) {
+                        LineRange identifierLineRange = child.lineRange();
+                        startCodeBodyPosition = new Position(identifierLineRange.endLine().line(),
+                                identifierLineRange.endLine().offset());
+                        break;
+                    }
+                }
+                LineRange typeLineRange = member.lineRange();
+                Position typeEndCodeBodyPosition = new Position(typeLineRange.endLine().line(),
+                        typeLineRange.endLine().offset());
+                return new NodePositions(startCodePosition, startCodeBodyPosition, typeEndCodeBodyPosition);
+        }
+        return new NodePositions(startCodePosition, null, null);
+    }
+
+    /**
+     * Returns the number of API documentation lines.
+     *
+     * @param member Node for which the doc lines are calculated
+     * @return Number of documentation lines
+     */
+    private static int getNumberOfDocLines(NonTerminalNode member) {
+        if (member.children().size() > 0 && member.children().get(0) instanceof MetadataNode) {
+            LineRange docLineRange = member.children().get(0).lineRange();
+            return docLineRange.endLine().line() - docLineRange.startLine().line() + 1;
+        }
+        return 0;
+    }
+
+    /**
+     * Checks if a give position is within a given range.
+     *
+     * @param start Start position of the range
+     * @param end End position of the range
+     * @param position Cursor position
+     * @param isEndInclusive If the range is included with end position or not
+     * @return If the position is within the range or not
+     */
+    private static boolean isWithinRange(Position start, Position end, Position position, boolean isEndInclusive) {
```

Comment:
Is it possible to use an offset based calculation here?

Reply:
Implementation is done based on the actual positions relative to the editor. If we are to do offset based calculation, we have to track the cumulative offset values up to these positions instead of the positions themselves. Are there any advantages of using offset over positions? IMO, using positions is more comprehensible to map with the editor and easy to debug whereas with offsets we will have to traverse through children to calculate the cumulative value up to a particular node. Wdyt?

Reply:
In ballerina syntax tree, positions are already captured as cumulative values. So there will be no issue with that.

Reply:
Updated the logic with offset calculations.